tanjun/cached_rest.py

```diff
@@ -0,0 +1,456 @@
+        self._invite_store.clear()
+        self._me_store.clear()
+        self._member_store.clear()
+        self._message_store.clear()
+        self._role_store.clear()
+        self._user_store.clear()
+
+    def gc(self) -> None:
+        self._application_store.gc()
+        self._channel_store.gc()
+        self._emoji_store.gc()
+        self._guild_store.gc()
+        self._invite_store.gc()
+        self._me_store.gc()
+        self._member_store.gc()
+        self._message_store.gc()
+        self._role_store.gc()
+        self._user_store.gc()
+
+    async def fetch_application(self) -> applications.Application:
+        if application := self._application_store.get():
+            with ERROR_MANAGER:
+                return await application
+
+            self._application_store.clear()  # type: ignore[unreachable]
+
+        fetched_application = await self._application_store.set(self._rest.rest.fetch_application())
+
+        if fetched_application.team:
+            self._user_store.update(fetched_application.team.members)
+
+        else:
+            self._user_store[fetched_application.owner.id] = fetched_application.owner
+
+        return fetched_application
+
+    async def fetch_channel(
+        self, channel: snowflakes.SnowflakeishOr[channels.PartialChannel], /
+    ) -> channels.PartialChannel:
+        channel = snowflakes.Snowflake(channel)
+        if cached_channel := self._channel_store.get(channel):
+            with ERROR_MANAGER:
+                return await cached_channel
+
+            self._channel_store.pop(channel, None)  # type: ignore[unreachable]
+
+        fetched_channel = await self._channel_store.set_future(channel, self._rest.rest.fetch_channel(channel))
+        if isinstance(fetched_channel, channels.DMChannel):
+            self._user_store[fetched_channel.recipient.id] = fetched_channel.recipient
+
+        elif isinstance(fetched_channel, channels.GroupDMChannel):
+            self._user_store.update(fetched_channel.recipients)
+
+        return fetched_channel
+
+    async def fetch_emoji(
+        self,
+        guild: snowflakes.SnowflakeishOr[guilds.PartialGuild],
+        emoji: snowflakes.SnowflakeishOr[emojis.CustomEmoji],
+        /,
+    ) -> emojis.KnownCustomEmoji:
+        emoji = snowflakes.Snowflake(emoji)
+        if isinstance(cached_emoji := self._emoji_store.get(emoji), asyncio.Future):
+            with ERROR_MANAGER:
+                return await cached_emoji
+
+            self._emoji_store.pop(emoji, None)  # type: ignore[unreachable]
+
+        elif cached_emoji:
+            return cached_emoji
+
+        fetched_emoji = await self._emoji_store.set_future(emoji, self._rest.rest.fetch_emoji(guild, emoji))
+
+        if fetched_emoji.user:
+            self._user_store[fetched_emoji.user.id] = fetched_emoji.user
+
+        return fetched_emoji
+
+    async def fetch_guild(self, guild: snowflakes.SnowflakeishOr[guilds.PartialGuild], /) -> guilds.RESTGuild:
```

Comment:
Should we type this as just guilds.Guild on the interface to allow some flexibility impl detail wise